src/bower_components/apiclient/connectionmanager.js

```diff
@@ -239,307 +396,543 @@ define(["events", "apiclient", "appStorage"], function(events, apiClientFactory,
+            var request = {
+                type: "POST",
+                url: getConnectUrl("pin/authenticate"),
+                data: {
+                    deviceId: pinInfo.DeviceId,
+                    pin: pinInfo.Pin
+                },
+                dataType: "json"
+            };
+            addAppInfoToConnectRequest(request);
+            return ajax(request);
+        }
+
+        console.log("Begin ConnectionManager constructor");
+        var self = this;
+        this._apiClients = [];
+        var connectUser;
+
+        self.connectUser = function () {
+            return connectUser;
+        };
+
+        self._minServerVersion = "3.2.33";
+
+        self.appVersion = function () {
+            return appVersion;
+        };
+
+        self.appName = function () {
+            return appName;
+        };
+
+        self.capabilities = function () {
+            return capabilities;
+        };
+
+        self.deviceId = function () {
+            return deviceId;
+        };
+
+        self.credentialProvider = function () {
+            return credentialProvider;
+        };
+
+        self.connectUserId = function () {
+            return credentialProvider.credentials().ConnectUserId;
+        };
+
+        self.connectToken = function () {
+            return credentialProvider.credentials().ConnectAccessToken;
+        };
+
+        self.getServerInfo = function (id) {
+            return credentialProvider.credentials().Servers.filter(function (s) {
+                return s.Id === id;
+            })[0];
+        };
+
+        self.getLastUsedServer = function () {
+            var servers = credentialProvider.credentials().Servers;
+            servers.sort(function (a, b) {
+                return (b.DateLastAccessed || 0) - (a.DateLastAccessed || 0);
+            });
+
+            if (servers.length) {
+                return servers[0];
+            }
+
+            return null;
+        };
+
+        self.addApiClient = function (apiClient) {
+            self._apiClients.push(apiClient);
+
+            var existingServers = credentialProvider.credentials().Servers.filter(function (s) {
+                return stringEqualsIgnoreCase(s.ManualAddress, apiClient.serverAddress()) || stringEqualsIgnoreCase(s.LocalAddress, apiClient.serverAddress()) || stringEqualsIgnoreCase(s.RemoteAddress, apiClient.serverAddress());
+            });
+            var existingServer = existingServers.length ? existingServers[0] : apiClient.serverInfo();
+
+            if (existingServer.DateLastAccessed = new Date().getTime(), existingServer.LastConnectionMode = ConnectionMode.Manual, existingServer.ManualAddress = apiClient.serverAddress(), apiClient.manualAddressOnly && (existingServer.manualAddressOnly = true), apiClient.serverInfo(existingServer), apiClient.onAuthenticated = function (instance, result) {
+                    return onAuthenticated(instance, result, {}, true);
+                }, !existingServers.length) {
```

Comment:
```suggestion
                }
                if (!existingServers.length) {
```
(if you accept my suggestion above, they go in pair)